packages/devtools_app/lib/src/provider/provider_list.dart

```diff
@@ -0,0 +1,213 @@
+// Copyright 2021 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
```

Comment:
how much functionality are you using from riverpod? Unless it is extensive it could make sense to avoid the dependency for now.

Reply:
It's used extensively
With it, the implementation of the instance viewer is fully declarative, and it takes care of the process of canceling pending evaluations whenever possible.

Reply:
Interesting. Can you point to the pattern that ensures evaluations are cancelled?

Reply:
It's the "autoDispose" functionality on providers (which `package:provider` sadly cannot support)
https://riverpod.dev/docs/concepts/modifiers/auto_dispose

It keeps track of the number of listeners on the associated provider (be it widgets or other providers (such as when chaining providers)). And when all listeners are removed, the provider state will automatically be destroyed.

Then providers can dispose a dispose method to cancel pending requests. That's the "Disposable" object which all providers use:

```dart
AutoDisposeProvider((ref) {
  final isAlive = Disposable();
  ref.onDispose(isAlive.dispose);
  
  
  // when the provider is destroyed, `isAlive` will be disposed, canceling the pending requests.
  await eval.eval('...', isAlive: isAlive);
})
```


This combines nicely with the collapse functionality too.
All object properties are associated with a unique provider. So when collapsing a property, its listener is removed, its state destroyed and pending requests canceled.

Reply:
This architecture also automatically deals with loading and error handling.

It relies on the fact that using `await` throws when the awaited future fails, purposefully letting the exception bubble. Then the provider will catch the exception, and use a union-type to force the UI to deal with loading/errors.

That's the `when(loading: ..., error: ..., data: ...)` you can see inside widgets.